fixes: #9737, add literal_processors for interval

indiVar0508 opened the pull request:
added literal processor for intervals for postgres and oracle db in their respective types module.
added tests for same in the dialects directory within test_types.py docs refered
postgres: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT oracle: https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements003.htm#i38598

- Didn't add YEARS to MONTH for oracle as it was not already supported 
- not sure on how to support precision for postgres interval seconds so added a fixme note there for now.

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
<!-- Describe your changes in detail -->

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.

**Have a nice day!**

indiVar0508 committed:
fixes: #9737, add literal_processors for interval

added literal processor for intervals for postgres and oracle db
in their respective types module.
added tests for same in the dialects directory within test_types.py
docs refered
postgres: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT
oracle: https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements003.htm#i38598

added typing to added literal_processor methods.
added for both methods added in postgres and oracle types module
added no_type_check as there was a conflict in overridden
`literal_processor` method for return type.

test: add Interval based test in test suite

added IntervalTest class in test_types file withing suite directory
refered & followed pattern as per Boolean, String and Date Based
Tests.

lelit commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -247,6 +247,90 @@ def as_generic(self, allow_nulltype: bool = False) -> sqltypes.Interval:
+                DAYS_IN_YEAR = 365.25  # Average days in a year
+
+                # Calculate years
+                years = seconds / (
```
Maybe divmod() is clearer/faster/on purpose here and below?

lelit commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -247,6 +247,90 @@ def as_generic(self, allow_nulltype: bool = False) -> sqltypes.Interval:
+                SECONDS_IN_MINUTE = 60
+                MINUTES_IN_HOUR = 60
+                HOURS_IN_DAY = 24
+                DAYS_IN_MONTH = 30.44  # Average days in a month
```
I'd probably go with _precise_ transformation, and avoid going to months/years that necessarily involves using such math with _averages_... AFAICT, PostgreSQL will do the "right" thing when computing, say, `'2024-02-29'::date + '1 year'::interval`, that gives `2025-02-28`, and it could be a surprise that the same expression in Python (where the math is done in fractions of seconds) gives a different result.

lelit replied:
To clarify what I meant, consider starting with `timedelta(seconds=1 * 365.25 * 24 * 60 * 60)`, it would surprise me if that would be rendered as `'1 year'::interval` : 
```
test=# select '2024-02-29'::date + '1 year'::interval;
      ?column?       
---------------------
 2025-02-28 00:00:00

test=# select '2024-02-29'::date + '31557600 seconds'::interval;
      ?column?       
---------------------
 2025-02-28 06:00:00
```

Even Python itself does not involve "imprecise" units:
```python
>>> timedelta(seconds=1 * 365.25 * 24 * 60 * 60)
datetime.timedelta(days=365, seconds=21600)
```

indiVar0508 replied:
Hi @lelit ,

Thanks for the feedback and explanation, i updated the approach, ditching the computing years and stuff bit, and directly using what timedelta gives.
saw this in [docs](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT), which should work ig?!
`Quantities of days, hours, minutes, and seconds can be specified without explicit unit markings. For example, '1 12:59:10' is read the same as '1 day 12 hours 59 min 10 sec'.`

indiVar0508 replied:
what i was also wondering was for how to control precision of seconds say if i have a interval as `0:00:3.23234234241234` and i want to round it to 3 decimal places for postgres i am not sure how to do it.
in the [docs](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-DATETIME-INPUT) it says to do `type [ (p) ] 'value'` but where will it fit in expression is something i am not able to figure out, can you point me in the right direction here.
Thanks

lelit replied:
I'm not sure I understand what you are asking: `select interval (3) '0:00:3.1289'` yields `00:00:03.129`, while `select (3) '0:00:3.23234234241234'` gives `00:00:03.232`, but you probably know that. If so, what do you mean with "_where will it fit in expression"_?

indiVar0508 replied:
Hi @lelit ,

`but you probably know that` - i am afraid and embarassed, i meant this only :"), i was trying like this `select interval '0:00:3.232323' second (3)` and it also worked for me (but doc said, type [ (p) ] 'value') so wasn't sure and confused, but thanks for helping!, i have updated my commit.

lelit replied:
You're welcome!

CaselIT commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -247,6 +247,14 @@ def as_generic(self, allow_nulltype: bool = False) -> sqltypes.Interval:
+    def literal_processor(self, dialect):
+        def process(value):
+            if value is not None:
+                value = f"'{str(value).replace(' days,', '')}'::interval"
```
seems we can also just do `f"'{value.total_seconds()}'::interval`

CaselIT replied:
`f"make_interval(secs=>{value.total_seconds()})"` works too

indiVar0508 replied:
Updated as per comments, went with `make_interval` method.
i was trying to write tests for execution of queries but was getting this error, have added for compile as of now, not able to figure out how to make tests point to specific db hosted in my local for testing pupose
`test.dialect.oracle.test_types.TypesTest' unsupported on any DB implementation 'sqlite(3.31.1)+pysqlite'`

CaselIT replied:
by default the query runs using sqlite, that can't execute oracle specific code, so skip are automatically added when tests are marked as "backend".

You can use oracle via docker with the tests, check the readme unit test: https://github.com/sqlalchemy/sqlalchemy/blob/main/README.unittests.rst#docker-configurations
or I can just make it run the pipeline on gerrit ;)

CaselIT replied:
mypy seems the only failure here. It logged
```
pep484: commands[0]> mypy ./lib/sqlalchemy
lib/sqlalchemy/dialects/postgresql/types.py:250: error: Function is missing a type annotation  [no-untyped-def]
lib/sqlalchemy/dialects/postgresql/types.py:251: error: Function is missing a type annotation  [no-untyped-def]
Found 2 errors in 1 file (checked 252 source files)
```

indiVar0508 replied:
Yeah i have added typing to method(s) but for some reason couldn't figure out how to resolve this, without modifying superclass returntype for now added no_type_check decorator ( saw that oracle had it for entire file, so added this for postgres in method only, and to keep it consistent also added it in oracle )
```py
error: Return type "Callable[[Optional[timedelta]], Optional[str]]" of "literal_processor" incompatible with return type "Optional[_LiteralProcessorType[timedelta]]" in supertype "TypeEngine"  [override]
```
added testcase for db specific backend testing also.

indiVar0508 commented on lib/sqlalchemy/dialects/oracle/types.py:
```diff
@@ -214,6 +214,25 @@ def as_generic(self, allow_nulltype=False):
+                    value += f"({self.day_precision})"
+                value += " TO SECOND"
+                if self.second_precision is not None:
+                    value += f"({self.second_precision})"
```
Hi @CaselIT ,

I saw this method but wasn't sure how to specify precisions in this, but i have updated according to comment because i am realising that for literal query they might not need precision ig or if they require query could be something like `select round_method_for_respective_db_dialect(interval, precision_value) ...`.
have updated the query and tests accordingly

CaselIT replied:
If I'm not mistaken precision or length is not taken into consideration when using literal formatting.
For example literal formatter of `String(10)` will not do `value_str[:10]`, so it should be fine to not have precision taken into consideration here.

@zzzeek do you have a general rule here?

indiVar0508 replied:
Hi,

Found a corner case on this, if i insert a timedelta of 100 days. then it fails `leading_precision is the number of digits in the leading field. Accepted values are 0 to 9. The default is 2.` so if i have a timedelta of 100 days which has 3 digits it fails.

```py
sqlalchemy.exc.DatabaseError: (cx_Oracle.DatabaseError) ORA-01873: the leading precision of the interval is too small
[SQL: INSERT INTO t (x) VALUES (NUMTODSINTERVAL(8640000.0, 'SECOND'))]
```

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 9a6f34506780d5953c36eb36a3ace8755b607a3d of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 9a6f34506780d5953c36eb36a3ace8755b607a3d: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 3019e9fb5a9e7ac94a013a07236719fb6fcee4ee of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 3019e9fb5a9e7ac94a013a07236719fb6fcee4ee added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

zzzeek reviewed (changes requested):
any chance we can get a suite test ?  here's where the date related tests start: https://github.com/sqlalchemy/sqlalchemy/blob/aea28a9607d3673c680a4ebe7b9fb672db745d0a/lib/sqlalchemy/testing/suite/test_types.py#L455    so this would be a new suite IntervalTest with some basic round trips and literal processor tests.

indiVar0508 commented:
Tried to add a few in [41528f6](https://github.com/sqlalchemy/sqlalchemy/pull/10383/commits/41528f6cb9d8c55e266731031fb8bb7cbc5f972a), you can review once if more are to be added or some changes needed to be done in these

CaselIT commented on lib/sqlalchemy/testing/suite/test_types.py:
```diff
@@ -452,6 +453,100 @@ def test_concatenate_clauselist(self, connection):
 
 
+class IntervalTest(_LiteralRoundTripFixture, fixtures.TestBase):
+    __requires__ = ("datetime",)
```
we probably need a new require here, called interval or similar

indiVar0508 replied:
Added requires you can review once went with name `datetime_interval`.

While adding only_on, just checked for other DB(s), where i think Interval is supported for arithmetic operations but it is not supported with sqla.

_MSSQL_ : https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/time-date-and-interval-functions?view=sql-server-ver16
_MySQL_ : https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html

CaselIT replied:
not sure if there is a reason why they are supported only on pg and oracle. @zzzeek do you remember why?

CaselIT commented on lib/sqlalchemy/testing/suite/test_types.py:
```diff
@@ -452,6 +453,100 @@ def test_concatenate_clauselist(self, connection):
+            ),
+        ]
+    )
+    def test_render_literal_interval(self, literal_round_trip):
```
combination seems unused, since you always use self.data

CaselIT commented on test/dialect/oracle/test_types.py:
```diff
@@ -323,6 +350,16 @@ def test_interval(self, metadata, connection):
             datetime.timedelta(days=35, seconds=5743),
         )
 
+    def test_interval_literal_processor(self, connection):
```
this test does not seem to do what it says. no literal processor is used.

maybe you wanted to use `parameter=literal(..., literal_execute=True)`

CaselIT commented on test/dialect/postgresql/test_types.py:
```diff
@@ -3180,6 +3180,16 @@ def test_interval_coercion_literal(self):
         expr = column("bar", postgresql.INTERVAL) == datetime.timedelta(days=1)
         eq_(expr.right.type._type_affinity, types.Interval)
 
+    def test_interval_literal_processor(self, connection):
```
same comment as in the oracle file

CaselIT reviewed (changes requested):
Thanks for updating!

Can you also add a changelog here https://github.com/sqlalchemy/sqlalchemy/tree/main/doc/build/changelog/unreleased_20 

feel free to add "pull request courtesy of your name"!

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision bf3a73dfd9d329779e12037ae62dc1032e76d0f6 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset bf3a73dfd9d329779e12037ae62dc1032e76d0f6 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

indiVar0508 commented:
Hi, 

It seemed some tests were failing, so fixed them some typo by me on comparing tuple vs datetime fixed that,
encountered a different issue in tests for now skipping them for oracle and added a Note on suite test.

It seems to be slight inconsistency in the behaviour of building query for datetime literal for postgres and oracle which causes this issue.
```py
import datetime
import sqlalchemy as sa
from sqlalchemy import create_engine


def main():
    engine = create_engine("postgresql+psycopg2://user_one:pass@127.0.0.1:5432/sqlalchemy_history_test")
    # engine = create_engine("oracle+cx_oracle://user_one:pass@127.0.0.1")
    conn = engine.connect()
    # This works!
    stmt = sa.select(sa.literal(datetime.datetime.now()) - sa.literal(datetime.timedelta(days=1), literal_execute=True))
    print(str(stmt.compile(dialect=engine.dialect, compile_kwargs={'literal_binds': True})))
    print(conn.execute(stmt).first())
    # This works!
    sql = 'SELECT :parameter - :parameter2 from dual' if engine.dialect.name == 'oracle' else 'SELECT :parameter - :parameter2'
    params = {'parameter': datetime.datetime.now(),'parameter2': datetime.timedelta(days=1)}
    stmt = sa.text(sql).bindparams(**params)
    print(str(stmt.compile(dialect=engine.dialect, compile_kwargs={'literal_binds': True})))
    print(conn.execute(stmt).first())
    # This Fails for postgres but works for oracle!

    # File "/github/sqlalchemy/lib/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    #   self.dialect.do_execute(
    # File "/github/sqlalchemy/lib/sqlalchemy/engine/default.py", line 921, in do_execute
    #   cursor.execute(statement, parameters)
    # sqlalchemy.exc.DataError: (psycopg2.errors.InvalidDatetimeFormat) invalid input syntax for type interval: "2023-09-29 12:53:23.380940"
    # LINE 1: SELECT '2023-09-29 12:53:23.380940' - make_interval(secs=>86...
    # [SQL: SELECT '2023-09-29 12:53:23.380940' - make_interval(secs=>86400.0) AS anon_1]
    # (Background on this error at: https://sqlalche.me/e/20/9h9h)

    stmt = sa.select(sa.literal(datetime.datetime.now(), literal_execute=True) - sa.literal(datetime.timedelta(days=1), literal_execute=True))
    print(str(stmt.compile(dialect=engine.dialect, compile_kwargs={'literal_binds': True})))
    print(conn.execute(stmt).first())

if __name__ == '__main__':
    main()
```

CaselIT commented:
Not too clear what's the issue.

pg wants the cast, IIRC sqlalchemy will add it if it knows the types of the columns.

What's the issue you get in the test where text is not used?

CaselIT commented:
@zzzeek do you remember if this case was meant to automatically cast in postgresql?

zzzeek commented:
well PG intervals have to be expressed in interval format, not as timestamps, so I know that much

CaselIT commented:
the issue does not seem to be on the interval, that uses `make_interval(secs=>86400.0)`, but in the datetime/time/date literal that's used for the subtraction. The test is of the type `select date_or_time_literal - the_interaval`

since this PR is for interval, we can probably just place a cast to the literal, so it renders `select cast(date_or_time_literal, the_type) - the_interval`. @indiVar0508 that should would and I don't think it will be rejected even in other databases.

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

if these are new tests then it's possible that subtracting a date literal and interval literal without casts just wont work.   i would look to see what the non-interval, date related tests are doing, if they have similar tests that are doing arithmetic + literal render + no casts.

I see in the failures a lot of trivial things like getting rows where scalars are expected etc., so I would just be getting as much of the test suite to pass as possible first, go for all green, then add trickier tests like arithmetic with literals.     I would want to review the code + basic test support first.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

indiVar0508 commented:
Hi @zzzeek ,

I fixed those tests in 4425fb5f10df567277f5fd4fb8c9a0d553229711, and tested with both DB(s) they should work now.
Was having above mentioned issue for postgres and a microsecond issue for oracle vs postgres so for now skipped the test for oracle and added note [here](https://github.com/sqlalchemy/sqlalchemy/blob/4425fb5f10df567277f5fd4fb8c9a0d553229711/lib/sqlalchemy/testing/suite/test_types.py#L498) and [here](https://github.com/sqlalchemy/sqlalchemy/blob/4425fb5f10df567277f5fd4fb8c9a0d553229711/lib/sqlalchemy/testing/suite/test_types.py#L544) related to microseconds.

Hi @CaselIT , 

`we can probably just place a cast to the literal`
Got it, i will get basic tests verified and then update to cast method if required also did you get a chance to review [this](https://github.com/sqlalchemy/sqlalchemy/pull/10383#discussion_r1341032759), this was another issue which i mentioned in one the comment thread. which seems to have issue for oracle for precision on there default method to make intervals.

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

this PR is a lot of effort so let me run it and see what the current issues are.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

indiVar0508 commented:
sure @zzzeek ,

i marked it as draft as #10513 is required for this ( to resolve Day to Second precision issue ) to make round_trip_tests work, 
rest all tests work fine i have updated some test in my local i can push, was waiting for other PR to finish first.

indiVar0508 commented:
Rebased on top of main

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -247,6 +250,17 @@ def as_generic(self, allow_nulltype: bool = False) -> sqltypes.Interval:
+        def process(value: Optional[dt.timedelta]) -> Optional[str]:
+            if value is not None:
+                return f"make_interval(secs=>{value.total_seconds()})"
+            return None
```
**mike bayer** (zzzeek) wrote:

this would return the string "NULL".  we dont typically implement or test None for the literal processors since I dont know that's too much of a real use case, ideally we'd address it all at once by having the compiler not actually call literal processors when the value is NULL in any case.  This does get into the issue of JSON "null" etc. but we have flags for that.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester commented on test/dialect/postgresql/test_types.py:
```diff
@@ -3180,6 +3180,31 @@ def test_interval_coercion_literal(self):
+            select(
+                literal(
+                    datetime.timedelta(days=1, minutes=3, seconds=4),
+                    literal_execute=True,
```
**mike bayer** (zzzeek) wrote:

this is the wrong form for parameters also

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

* test/dialect/postgresql/test_types.py (line 6275): this should likely be merged above

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

this is good now thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873 has been **merged**. Congratulations! :)

sqla-tester commented on lib/sqlalchemy/dialects/oracle/types.py:
```diff
@@ -214,6 +226,21 @@ def as_generic(self, allow_nulltype=False):
         )
 
+    @property
+    def python_type(self) -> Type[dt.timedelta]:
```
**Federico Caselli** (CaselIT) wrote:

why this decorator?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

somehow i just didnt see that.  will remove

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/types.py:
```diff
@@ -247,6 +250,17 @@ def as_generic(self, allow_nulltype: bool = False) -> sqltypes.Interval:
     def python_type(self) -> Type[dt.timedelta]:
         return dt.timedelta
 
+    @no_type_check
```
**Federico Caselli** (CaselIT) wrote:

here too

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

sqla-tester reviewed (commented):
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4873

indiVar0508 commented on lib/sqlalchemy/testing/suite/test_types.py:
```diff
@@ -452,6 +453,123 @@ def test_concatenate_clauselist(self, connection):
+        )
+
+    @testing.combinations(
+        (
```
Hi, 
this is minor but just wanted to bring it up here, 
this testcase was failing with for days > 99days {Precision of more than 2 digits } with oracle DB for round trip test, due #10509, ig we don't have this testcase now with that scenario, in PR for #10513 only compilation tests are added not the round trip as literal processor thing which was being added in this PR.

indiVar0508 replied:
Yeah sure, 

So at that time Interval Parameter were ineffective for oracleDB due to issue-10509, 
But my Interval param were `day_precision=9` and `second_precision=9`, 

Now this will work since PR-10513 is merged.

the testcase was this

```py
    @testing.combinations(
        (
            datetime.timedelta(days=103, seconds=4),
            datetime.timedelta(days=103, seconds=4),
        ),
        argnames="input_, compare_",
    )
    def test_render_literal_interval(
        self, literal_round_trip, input_, compare_
    ):
        literal_round_trip(Interval(), [input_], [compare_])
```

i remvoed the 3 digit day test case to make CI work for simple testcases in this PR.

zzzeek replied:
OK that is restored in  https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4939 if you want to review

indiVar0508 replied:
Yep looks good!
